protocol/protocol_test.go

```diff
@@ -0,0 +1,186 @@
+
+}
+
+func runStory(t *testing.T, conn io.ReadWriter, steps []pgstories.Step) error {
+	frontend, err := pgproto3.NewFrontend(conn, conn)
+	if err != nil {
+		return err
+	}
+
+	story := &pgstories.Story{
+		Steps:    steps,
+		Frontend: frontend,
+	}
+
+	sigKill := make(chan interface{})
+	timer := time.NewTimer(time.Second * 2)
+	go func() {
+		<-timer.C
+		sigKill <- fmt.Errorf("timeout")
+	}()
+
+	err = story.Run(t, sigKill)
+	if err != nil {
+		timer.Stop()
+	}
+	return err
+}
+
+func TestProtocol_Read(t *testing.T) {
+
+	t.Run("standard message flow", func(t *testing.T) {
+
+		f, b := net.Pipe()
+
+		frontend, err := pgproto3.NewFrontend(f, f)
+		require.NoError(t, err)
+
+		p := NewProtocol(b, b)
+		p.initialized = true
+
+		msg := make(chan Message)
+		go func() {
+			m, err := p.Read()
+			require.NoError(t, err)
+
+			msg <- m
+		}()
+
+		m, err := frontend.Receive()
+		require.NoError(t, err)
+		require.IsType(t, &pgproto3.ReadyForQuery{}, m, "expected protocol to send ReadyForQuery message")
+
+		_, err = f.Write([]byte{'Q', 0, 0, 0, 4})
+		require.NoError(t, err)
+
+		res := <-msg
+		require.Equalf(t, byte('Q'), res.Type(), "expected protocol to identify sent message as type 'Q'. actual: %c", res.Type())
+
+		require.Nil(t, p.transaction, "expected protocol not to start transaction")
+
+	})
+
+	t.Run("extended query message flow", func(t *testing.T) {
+
+		t.Run("starts transaction", func(t *testing.T) {
+
+			f, b := net.Pipe()
+
+			p := NewProtocol(b, b)
+			p.initialized = true
+
+			go func() {
+				for {
+					_, err := p.Read()
+					require.NoError(t, err)
+				}
+			}()
+
+			err := runStory(t, f, []pgstories.Step{
+				&pgstories.Response{BackendMessage: &pgproto3.ReadyForQuery{}},
+				&pgstories.Command{FrontendMessage: &pgproto3.Parse{}},
+				&pgstories.Command{FrontendMessage: &pgproto3.Bind{}},
+			})
+			require.NoError(t, err)
+
+			require.NotNil(t, p.transaction, "expected protocol to start transaction")
+
+		})
+
+		t.Run("ends transaction", func(t *testing.T) {
+
+			f, b := net.Pipe()
+
+			p := NewProtocol(b, b)
+			p.initialized = true
+
+			go func() {
+				for {
+					m, err := p.Read()
+					require.NoError(t, err)
+
+					switch m.Type() {
+					case Parse:
+						err = p.Write(ParseComplete)
+					case Bind:
+						err = p.Write(BindComplete)
+					}
+					require.NoError(t, err)
+				}
+			}()
+
+			err := runStory(t, f, []pgstories.Step{
+				&pgstories.Response{BackendMessage: &pgproto3.ReadyForQuery{}},
+				&pgstories.Command{FrontendMessage: &pgproto3.Parse{}},
+				&pgstories.Command{FrontendMessage: &pgproto3.Bind{}},
+				&pgstories.Command{FrontendMessage: &pgproto3.Sync{}},
+				&pgstories.Response{BackendMessage: &pgproto3.ParseComplete{}},
+				&pgstories.Response{BackendMessage: &pgproto3.BindComplete{}},
+				&pgstories.Response{BackendMessage: &pgproto3.ReadyForQuery{}},
+			})
+
+			require.NoError(t, err)
+
+			require.Nil(t, p.transaction, "expected protocol not to end transaction")
```

Comment:
If `p.transaction` should be `nil`, doesn't it mean that we expect protocol to end transaction?